protocol/test-sequencing.md

```diff
@@ -0,0 +1,216 @@
+    - `test_includeByHash(jobid, txHash)`: include specific tx from tx-pool
+    - `test_includeRaw(jobid, rawTx)`: include specific given tx
+    - `test_cancel(jobid)`: stop building a block
+    - `test_seal(jobid)`: complete building a block
+    - `test_auto(active)`: automatically repeat block-building
+    - `test_inspect(jobid) -> partial block`: inspect thus-far built block
+
+These RPC calls can then be used by tests to include the transactions in the desired order, in the specific blocks,
+to create the right test scenario. E.g. coordinating a cyclic transaction dependency,
+or including an otherwise invalid transaction (invalid executing message).
+
+#### L1 mode
+
+L1-block building shares a lot of code with L2, and can be done for testing as well.
+No system-transactions or formatted extra attributes like the Holocene `extraData` need to be inserted.
+
+One way to use this for L1 would be as "mock": to make the chain of blocks proactively.
+This can support shadow-forking (since L1 consensus is not followed), and avoid the need to run a beacon-node at all.
+
+Another way might be to intercept the engine-API `engine_forkchoiceUpdated(fcState, attributes)`,
+which can then queue the block-building work and complete it when there is a matching test block-building job.
+This is similar to the sidecar design in OP-Stack.
+However, this then makes the op-node lead, and prevents reorgs from being initiated by the test tool.
+
+#### Compared to op-wheel
+
+`op-wheel` may be time to deprecate.
+It consists of `db` (patch geth DB contents) and `engine` utils.
+
+The forkchoice / reorg CLI tools are still used,
+but may be possible to move to op-node or op-geth.
+
+The `engine` CLI tools are mostly one-off commands,
+not suited to build blocks with particular transactions, nor a kurtosis standalone service.
+
+There is an `engine auto` command for automatic block-building of L1 blocks (no L2 system deposits)
+on an interval using the engine-API.
+And it provides none of the testing or production sequencer functionality.
+
+### Production
+
+What if we want to make the test sequencer a production sequencer?
+
+The following integration work is needed:
+- Use `auto` sequencing mode by default
+- Detect when op-node is deriving conflicting blocks, and automatically back-off.
+    - When the last seen parent-block is not the tip of the chain.
+    - When `unsafe == safe`, aka likely syncing from L1 data.
+- Host the same sequencer admin RPC as op-node, to hande sequencer rotation:
+    - `admin_startSequencer`: start
+    - `admin_stopSequencer`: stop
+    - `admin_sequencerActive`: sequencer state
+    - `admin_postUnsafePayload`: publish a block, for failover
+    - `admin_overrideLeader`: used by op-conductor
+    - `admin_conductorEnabled`: used by op-conductor
+- Connect to geth P2P, to read/write the shared tx-pool.
+- Support the DA-throttling RPC.
+  Or better, implement proper multi-block DA congestion safe-guards.
```

Comment:
Building and testing this out for production is a pretty significant scope - particularly the need to fully support op-conductor. I suspect we should focus on testing needs now - our test code should still be high quality and anything complex (like this) should have tests so we know our tests are doing what we expect anyway.  Then we can separately look at product priorities and determine if we should build a separate block builder service for production (and exactly what it should support).

That said, would probably be good to get product input on what block building improvements are being considered/asked for etc so we're at least aware. (cc @elihaims ?)

Reply:
+1, agreed on priorities. After delivering the test scope however, I do strongly believe we need to revisit sequencing for production interop deployments, as part of this tool or not, given all the requirements and DoS risks that we discuss during testing calls.